packages/frontend/src/models/CanonicalBridge.ts

```diff
@@ -0,0 +1,341 @@
+        const l2TokenAddress = this.getL2CanonicalTokenAddress(this.tokenSymbol, this.chain)
+        if (!l2TokenAddress) {
+          throw new Error(
+            `token "${this.tokenSymbol}" on chain "${this.chain.slug}" is unsupported`
+          )
+        }
+
+        if (this.tokenSymbol === Token.DAI) {
+          const bridge = L1OptimismDaiTokenBridge__factory.connect(
+            l1CanonicalBridge.address,
+            this.signer
+          )
+          return bridge.populateTransaction.depositERC20(
+            l1CanonicalToken.address,
+            l2TokenAddress,
+            amount,
+            BigNumber.from('1920000'),
+            '0x',
+            {
+              from,
+              gasLimit: 500e3,
+            }
+          )
+        }
+
+        const bridge = L1OptimismTokenBridge__factory.connect(
+          l1CanonicalBridge.address,
+          this.signer
+        )
+        return bridge.populateTransaction.deposit(
+          l1CanonicalToken.address,
+          l2TokenAddress,
+          signerAddress,
+          payload,
+          { from, gasLimit: 500e3 }
+        )
+      }
+
+      case ChainSlug.Arbitrum: {
+        const ethSigner = await this.getSignerOrProvider(Chain.Ethereum, this.signer)
+        const arbProvider = await this.getSignerOrProvider(Chain.Arbitrum)
+        const arbSigner = (arbProvider as any).getUncheckedSigner(from)
+        const arbBridge = await Bridge.init(ethSigner as JsonRpcSigner, arbSigner)
+
+        let bridge: ArbitrumInbox | L1ArbitrumDaiGateway = L1ArbitrumDaiGateway__factory.connect(
+          l1CanonicalBridge.address,
+          this.signer
+        )
+
+        if (this.tokenSymbol === Token.ETH) {
+          bridge = await this.getContract(
+            ArbitrumInbox__factory,
+            l1CanonicalBridge.address,
+            this.signer
+          )
+          const maxSubmissionCost = BigNumber.from(amount).add(1000e4)
+          return ((bridge as ArbitrumInbox).populateTransaction.depositEth as any)(
+            BigNumber.from(maxSubmissionCost),
+            {
+              value: amount,
+              from,
+            }
+          )
+        }
+
+        const depositInputParams = {
+          erc20L1Address: l1CanonicalToken.address,
+          amount: BigNumber.from(amount),
+          destinationAddress: recipient,
+        }
+        const depositTxParams = await arbBridge.getDepositTxParams(depositInputParams)
+        const gasNeeded = await arbBridge.estimateGasDeposit(depositTxParams)
+        const { maxFeePerGas, gasPrice } = await arbBridge.l1Provider.getFeeData()
+        const price = maxFeePerGas || gasPrice
+        const walletBal = await arbBridge.l1Provider.getBalance(from)
+
+        if (!price) {
+          console.log('Warning: could not get gas price estimate; will try depositing anyway')
+        } else {
+          const fee = price.mul(gasNeeded)
+          if (fee.gt(walletBal)) {
+            console.log(
```

Comment:
does metamask throw error in this case?

Reply:
that is a good question. i haven't run into this issue yet, and this is copy/pasta from an [arb-ts example](https://github.com/OffchainLabs/arbitrum/blob/338fa2da63c290faeaa27b848f1d93d60038aa09/packages/arb-ts/scripts/deployStandard.ts#L121)